src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs

```diff
@@ -0,0 +1,97 @@
+﻿namespace NServiceBus.Gateway
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using NServiceBus.DelayedDelivery;
+    using NServiceBus.DeliveryConstraints;
+    using NServiceBus.Logging;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Transports;
+
+    class RetryFailedGatewayMessagesBehavior : ForkConnector<ITransportReceiveContext, IRoutingContext>
+    {
+        public RetryFailedGatewayMessagesBehavior(string localAddress, Func<IncomingMessage, Exception, int, TimeSpan> retryPolicy)
+        {
+            this.localAddress = localAddress;
+            this.retryPolicy = retryPolicy;
+        }
+
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next, Func<IRoutingContext, Task> fork)
+        {
+            try
+            {
+                await next().ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                var sentForRetry = await SendForRetry(context, exception, fork).ConfigureAwait(false);
+                if (!sentForRetry)
+                {
+                    throw;
+                }
+            }
+        }
+
+        private async Task<bool> SendForRetry(ITransportReceiveContext context, Exception exception, Func<IRoutingContext, Task> fork)
+        {
+            var message = context.Message;
+
+            var currentRetry = GetCurrentRetry(message);
+
+            var timeIncrease = retryPolicy(message, exception, currentRetry);
+
+            if (timeIncrease <= TimeSpan.Zero)
```

Comment:
in the core, `TimeSpan.Zero` means immediate retry in the custom policy, where `TimeSpan.MinValue` means "do not retry". I think it would make sense to align these?